### ISSUE TITLE
Alter context before creating project

### ISSUE THREAD

**Issue by italomaia** [audreyr/cookiecutter#364]
It would be nice if cookiecutter could run a function into it's context before creating the project. Example:

```
project
-- {{ cookiecutter.repo_name }}/ ...
-- context.py
```

**context.py**
```
def update_context(context):
    # do stuff to context
```

**Comment by italomaia** [comment]
It could also, instead, allow cookiecutter to be initialized from a custom script. That could allow for a lot of flexibility.

**Comment by slothyrulez** [comment]
+1 on this.

Is not obvious how to inject dynamic context, ej. use the documentation [snippet] (http://cookiecutter.readthedocs.org/en/latest/advanced_usage.html#example-injecting-a-timestamp) on the pre_gen_project hook may cause an 'infinite' prompt for values loop.

**Comment by DanEEStar** [comment]
I am also interested in that.

I couldn't find a way to dynamically add a context variable. So that I can add a secret_key context variable like mentioned in https://github.com/pydanny/cookiecutter-django/issues/3.

**Comment by saritasa** [comment]
+1 on this that will allow us to ask for additional questions, like is use_pusher="Y" then ask addition api keys to be generated inside config file.

**Comment by econchick** [comment]
I'd also like to have this functionality. I'd like to be able to edit/add to the template context either as it's being processed (like how [template variables are accessible](http://cookiecutter.readthedocs.org/en/latest/advanced_usage.html#templates-in-context-values) within the `cookiecutter.json` file itself right after the user is prompted), or perhaps a hook before the project is generated.  

For instance, the current problem (and have other similar problems) I'd like to solve is to dynamically create the version number depending on what the user inputs.  An example json file:

```js
{
    "dev_status": ["planning", "pre-alpha", "alpha", "beta", "stable", "mature", "inactive"]
    // then interpolate the project version based on the selection from above
    // currently I can do this:
    "version": "{% if cookiecutter.dev_status == 'alpha' -%}0.0.1a1{% else -%}0.0.1{% endif -%}",
}
```

If I were to do the full if/then for the version, it'd get quite long and unreadable. I'd prefer to do this a bit easier and outside of the JSON file, maybe like a jinja macro or something. 

(/me waves hello to @audreyr and @pydanny )

**Comment by goanpeca** [comment]
Yes, something like this would be nice. Is there any other way to define global context variables?

**Comment by rmedaer** [comment]
Hi guys,

Another way to alter context would be from **pre** hook script. The `pre_gen_project` could write a JSON object on stdout. Then Cookiecutter could read this object and (deep) merge it to input context.

Btw passing through stdin/stdout could be also an improvement in `post_gen_project` (instead of templating using Jinja).

This mechanism is used by softwares like Ansible to execute modules, hooks, plugins...

In this case, the context should be: 
```python
context = deep_merge(deep_merge(extra_context, user_input_context), pre_gen_project_context)
```

**Comment by h-mayorquin** [comment]
I wish this was in the tutorials at the `readthedocs`.

**Comment by att14** [comment]
If you are already writing a python function, you could call `cookiecutter` from python instead. https://cookiecutter.readthedocs.io/en/1.7.2/advanced/injecting_context.html

**Comment by connewm** [comment]
Does anyone know if this has been resolved? From what I can tell it would be relatively easy to implement and I would be happy to do it. 

As @att14 stated we can just inject context and re-call cookiecutter. While it is possible to do this seperately within the hook, I see benefit to adding this functionality to the module. Forgive me if I am mistaken and this issue has already been resolved.
